### PR TITLE
avoid freeze with multiple layers locator search

### PR DESCRIPTION
**Pull request by 3nids**
if the project had many searchable layers, a freeze occured when running the all features locator filter
the feature iterator was created in prepare (main thread) and the maximum number of connection was reached

### CODE REVIEW

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 1]
```diff
@@ -364,13 +366,14 @@ void QgsAllLayersFeaturesLocatorFilter::prepare( const QString &string, const Qg
...
     req.setLimit( 30 );
 
-    PreparedLayer preparedLayer;
-    preparedLayer.expression = expression;
-    preparedLayer.context = context;
-    preparedLayer.layerId = layer->id();
-    preparedLayer.layerName = layer->name();
-    preparedLayer.iterator =  layer->getFeatures( req );
-    preparedLayer.layerIcon = QgsMapLayerModel::iconForLayer( layer );
+    PreparedLayer *preparedLayer = new PreparedLayer();
```
Comment (nyalldawson) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp]: This leaks, right?

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 2]
```diff
@@ -383,23 +386,24 @@ void QgsAllLayersFeaturesLocatorFilter::fetchResults( const QString &string, con
   QgsFeature f;
 
   // we cannot used const loop since iterator::nextFeature is not const
-  for ( PreparedLayer preparedLayer : mPreparedLayers )
+  for ( auto preparedLayer : mPreparedLayers )
```
Comment (m-kuhn) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp]: qgis::as_const( mPreparedLayers )

File: src/app/locator/qgsinbuiltlocatorfilters.cpp (hunk 1 above)
Comment (m-kuhn) [reply src/app/locator/qgsinbuiltlocatorfilters.cpp]: Either make a `QList<std::shared_ptr<PreparedLayer> > mPreparedLayers;` or a `std::shared_ptr<QgsVectorLayerFeatureSource>`

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 3]
```diff
@@ -364,15 +367,16 @@ void QgsAllLayersFeaturesLocatorFilter::prepare( const QString &string, const Qg
...
+    preparedLayer->expression = expression;
+    preparedLayer->context = context;
+    preparedLayer->layerId = layer->id();
+    preparedLayer->layerName = layer->name();
+    preparedLayer->featureSource.reset( new QgsVectorLayerFeatureSource( layer ) );
+    preparedLayer->request = req;
+    preparedLayer->layerIcon = QgsMapLayerModel::iconForLayer( layer );
 
-    mPreparedLayers.append( preparedLayer );
+    mPreparedLayers.append( std::shared_ptr<PreparedLayer>( preparedLayer ) );
```
Comment (m-kuhn) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp]: I would initialize that already on line 370, then memory management is exception proof straight from the beginning.

**Review by m-kuhn — APPROVED** [review]

**Pull request merged**